File: ztm-ds-algo/6-binary-trees/q15-max-depth-of-binary-tree/q15-sol.cpp
// LeetCode 104.
// https://leetcode.com/problems/maximum-depth-of-binary-tree/

/**
 * Definition for a binary tree node.
 */
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (!root) return 0;
        
        int depth = 1;
        depth += max(maxDepth(root->left), maxDepth(root->right));
        return depth;
    }
};

// time: O(n)
// space: O(n)   binary tree worst case

File: ztm-ds-algo/6-binary-trees/q15-max-depth-of-binary-tree/q15-sol.py
# // LeetCode 104.
# // https://leetcode.com/problems/maximum-depth-of-binary-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root):
        if root is None:
            return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))

# // time: O(n)
# // space: O(n)   binary tree worst case